File: sdk/typescript/src/clients/agents.ts
/**
 * Agents resource client.
 *
 * Design Reference: DR-5
 * Requirements: 6.1, 6.2, 6.3, 6.4
 */

import type { HTTPTransport } from '../transport.js';
import type { Agent, AgentProfile, Reputation } from '../types/index.js';

/**
 * Parse ISO date string to Date object.
 */
function parseDate(dateStr: string): Date {
  return new Date(dateStr.replace(/Z$/, '+00:00'));
}

/**
 * Client for agent-related operations.
 *
 * Design Reference: DR-5
 */
export class AgentsClient {
  private transport: HTTPTransport;

  /**
   * Initialize the agents client.
   *
   * @param transport - HTTP transport for making requests
   */
  constructor(transport: HTTPTransport) {
    this.transport = transport;
  }

  /**
   * Register a new agent.
   *
   * This is an unsigned request - no authentication required.
   *
   * @param agentName - Display name for the agent
   * @param publicKey - Public key in format "ed25519:base64..." or "ecdsa:base64..."
   * @param capabilities - Optional list of agent capabilities
   * @returns Agent object with agentId, agentName, and createdAt
   * @throws ValidationError if agentName or publicKey is invalid
   * @throws ConflictError if agentName already exists
   *
   * Requirements: 6.1, 6.2
   */
  async register(
    agentName: string,
    publicKey: string,
    capabilities?: string[]
  ): Promise<Agent> {
    const body: Record<string, unknown> = {
      agentName,
      publicKey,
    };
    if (capabilities) {
      body.capabilities = capabilities;
    }

    const response = await this.transport.unsignedRequest(
      'POST',
      '/v1/agents/register',
      undefined,
      body
    );

    const data = (response.data ?? {}) as Record<string, unknown>;
    return {
      agentId: data.agentId as string,
      agentName: data.agentName as string,
      createdAt: parseDate(data.createdAt as string),
    };
  }

  /**
   * Get agent profile.
   *
   * @param agentId - The unique agent identifier
   * @returns AgentProfile with agent details and capabilities
   * @throws NotFoundError if agent not found
   *
   * Requirements: 6.3
   */
  async get(agentId: string): Promise<AgentProfile> {
    const response = await this.transport.unsignedRequest(
      'GET',
      `/v1/agents/${agentId}`
    );

    const data = (response.data ?? {}) as Record<string, unknown>;
    return {
      agentId: data.agentId as string,
      agentName: data.agentName as string,
      capabilities: (data.capabilities as string[]) ?? [],
      createdAt: parseDate(data.createdAt as string),
    };
  }

  /**
   * Get agent reputation score.
   *
   * @param agentId - The unique agent identifier
   * @returns Reputation with score (0.0 to 1.0) and updatedAt
   * @throws NotFoundError if agent not found
   *
   * Requirements: 6.4
   */
  async getReputation(agentId: string): Promise<Reputation> {
    const response = await this.transport.unsignedRequest(
      'GET',
      `/v1/agents/${agentId}/reputation`
    );

    const data = (response.data ?? {}) as Record<string, unknown>;
    // Handle both camelCase and snake_case from backend
    const agentIdVal = (data.agentId ?? data.agent_id) as string;
    const updatedAtVal = (data.updatedAt ?? data.updated_at) as string;
    return {
      agentId: agentIdVal,
      score: data.score as number,
      updatedAt: parseDate(updatedAtVal),
    };
  }
}

File: sdk/rust/src/clients/agents.rs
//! Agents resource client.
//!
//! Design Reference: DR-5
//! Requirements: 6.1, 6.2, 6.3, 6.4

use std::sync::Arc;

use serde_json::Value;

use crate::error::Error;
use crate::transport::HttpTransport;
use crate::types::{Agent, AgentProfile, Reputation};

/// Client for agent-related operations.
pub struct AgentsClient {
    transport: Arc<HttpTransport>,
}

impl AgentsClient {
    /// Create a new agents client.
    pub fn new(transport: Arc<HttpTransport>) -> Self {
        Self { transport }
    }

    /// Register a new agent.
    ///
    /// This is an unsigned request - no authentication required.
    ///
    /// # Arguments
    ///
    /// * `agent_name` - Display name for the agent
    /// * `public_key` - Public key in format "ed25519:base64..." or "ecdsa:base64..."
    /// * `capabilities` - Optional list of agent capabilities
    ///
    /// # Returns
    ///
    /// Agent object with agent_id, agent_name, and created_at
    ///
    /// # Errors
    ///
    /// Returns an error if registration fails.
    ///
    /// Requirements: 6.1, 6.2
    pub async fn register(
        &self,
        agent_name: &str,
        public_key: &str,
        capabilities: Option<Vec<String>>,
    ) -> Result<Agent, Error> {
        let mut body = serde_json::json!({
            "agentName": agent_name,
            "publicKey": public_key,
        });

        if let Some(caps) = capabilities {
            body["capabilities"] = serde_json::json!(caps);
        }

        let response: Value = self
            .transport
            .unsigned_request("POST", "/v1/agents/register", None, Some(&body))
            .await?;

        let data = response
            .get("data")
            .ok_or_else(|| Error::Http("Missing data in response".to_string()))?;

        serde_json::from_value(data.clone()).map_err(Error::from)
    }

    /// Get agent profile.
    ///
    /// # Arguments
    ///
    /// * `agent_id` - The unique agent identifier
    ///
    /// # Returns
    ///
    /// AgentProfile with agent details and capabilities
    ///
    /// # Errors
    ///
    /// Returns an error if the agent is not found.
    ///
    /// Requirements: 6.3
    pub async fn get(&self, agent_id: &str) -> Result<AgentProfile, Error> {
        let response: Value = self
            .transport
            .unsigned_request::<Value>("GET", &format!("/v1/agents/{agent_id}"), None, None::<&()>)
            .await?;

        let data = response
            .get("data")
            .ok_or_else(|| Error::Http("Missing data in response".to_string()))?;

        serde_json::from_value(data.clone()).map_err(Error::from)
    }

    /// Get agent reputation score.
    ///
    /// # Arguments
    ///
    /// * `agent_id` - The unique agent identifier
    ///
    /// # Returns
    ///
    /// Reputation with score (0.0 to 1.0) and updated_at
    ///
    /// # Errors
    ///
    /// Returns an error if the agent is not found.
    ///
    /// Requirements: 6.4
    pub async fn get_reputation(&self, agent_id: &str) -> Result<Reputation, Error> {
        let response: Value = self
            .transport
            .unsigned_request::<Value>(
                "GET",
                &format!("/v1/agents/{agent_id}/reputation"),
                None,
                None::<&()>,
            )
            .await?;

        let data = response
            .get("data")
            .ok_or_else(|| Error::Http("Missing data in response".to_string()))?;

        serde_json::from_value(data.clone()).map_err(Error::from)
    }
}
